sql: add typecasts to expressions shown in pg_catalog

rafiss opened the pull request:
fixes #53942 

also includes three small cleanup commits:
colexec: reduce code duplication in mergejoiner
descpb: fix comment about formatting expressions
sql: consolidate FormatExprForDisplay functions

pg_attrdef.adbin (column defaults) and pg_constraint.condef (column
constraints) are supposed to format the expression using typecasts for
non-numeric constants. This was previously being done for string
constants in pg_attrdef, but a recent change broke that functionality.

Release note (sql change): SQL expressions that are shown in pg_catalog
columns and related functions now will be formatted with a typecast for
non-numeric constants.

Release justification: low-risk change that improves compatibility, and
also fixes a regression that started in this release.

rafiss committed:
sql: add typecasts to expressions shown in pg_catalog

pg_attrdef.adbin (column defaults) and pg_constraint.condef (column
constraints) are supposed to format the expression using typecasts for
non-numeric constants. This was previously being done for string
constants in pg_attrdef, but a recent change broke that functionality.

Release note (sql change): SQL expressions that are shown in pg_catalog
columns and related functions now will be formatted with a typecast for
non-numeric constants.
Release justification: low-risk change that improves compatibility, and
also fixes a regression that started in this release.

rafiss committed:
colexec: reduce code duplication in mergejoiner

Use the new IsNumeric function.

Release note: None
Release justification: low-risk refactoring

rafiss committed:
descpb: fix comment about formatting expressions

It was referring to a function that no longer exists.

Release note: None
Release justification: comment-only change

mgartner commented on pkg/sql/schemaexpr/expr.go:
```diff
@@ -136,6 +136,21 @@ func FormatExprForDisplayWithoutTypeAnnotations(
+// and does add typecasts for non-numeric constants. This is the format that
+// should be used for displaying expressions in pg_catalog, in order to achieve
+// compatibility with the pg_catalog from PostgreSQL.
+func FormatExprForDisplayInPGCatalog(
```
[nit] Create a format option enum and combine these three `FormatExprForDisplay...` functions into one. They're all very similar.

rafiss replied:
good call

mgartner commented on pkg/sql/sem/tree/format.go:
```diff
@@ -365,7 +366,7 @@ func (ctx *FmtCtx) FormatNode(n NodeFormatter) {
 		}
 	}
-	if f.HasFlags(fmtDisambiguateDatumTypes) {
+	if f.HasFlags(fmtDisambiguateDatumTypes) || f.HasFlags(FmtPGCatalog) {
```
[nit] `if f.HasFlags(fmtDisambiguateDatumTypes | FmtPGCatalog) {`

rafiss replied:
done

rafiss replied:
Turns out `HasFlags` checks if _all_ of the flags are set, so I wrote a new `HasAnyFlags` function and used that.

mgartner reviewed (approved):
LGTM, just 2 optional changes. Happy to take another look if you decide to make the changes.

rafiss committed:
sql: consolidate FormatExprForDisplay functions

There were 3 functions doing similar things.

Release note: None
Release justification: low-risk refactoring

rafiss commented:
This is ready for another look

mgartner commented on pkg/sql/schemaexpr/expr.go:
```diff
@@ -111,29 +111,21 @@ func ExtractColumnIDs(desc catalog.TableDescriptor, rootExpr tree.Expr) (TableCo
 
-// FormatExprForDisplay formats a schema expression string for display by adding
-// type annotations and resolving user defined types.
+// FormatExprForDisplayInPGCatalog formats a schema expression string for
```
-> `FormatExprForDisplay`

mgartner commented on pkg/sql/pg_catalog.go:
```diff
@@ -414,19 +413,11 @@ CREATE TABLE pg_catalog.pg_attrdef (
 			}
-			var defSrc *tree.DString
-			expr, err := parser.ParseExpr(*column.DefaultExpr)
+			displayExpr, err := schemaexpr.FormatExprForDisplayInPGCatalog(ctx, table, *column.DefaultExpr, &p.semaCtx)
```
-> `FormatExprForDisplay(..., tree.FmtPGCatalog)`

mgartner commented on pkg/sql/pg_catalog.go:
```diff
@@ -978,7 +969,7 @@ func populateTableConstraints(
 				return err
 			}
-			displayExpr, err := schemaexpr.FormatExprForDisplayWithoutTypeAnnotations(ctx, table, con.Details, &p.semaCtx)
+			displayExpr, err := schemaexpr.FormatExprForDisplayInPGCatalog(ctx, table, con.Details, &p.semaCtx)
```
-> `FormatExprForDisplay(..., tree.FmtPGCatalog)`

mgartner reviewed (commented):
Nice! Looks great, you just missed 3 things.

rafiss commented:
Thanks! seems like I rushed too much

mgartner reviewed (approved):
```
 _     ____ _____ __  __ 
| |   / ___|_   _|  \/  |
| |  | |  _  | | | |\/| |
| |__| |_| | | | | |  | |
|_____\____| |_| |_|  |_|
```

rafiss commented:
Turns out `HasFlags` checks if all of the flags are set, so I wrote a new `HasAnyFlags` function and used that.

rafiss commented:
bors r=mgartner
